core/chains/evm/txmgr/finalizer.go

```diff
@@ -0,0 +1,299 @@
+		// The tx store query ensures transactions have receipts but leaving this check here for a belts and braces approach
+		if receipt.Receipt.IsZero() || receipt.Receipt.IsUnmined() {
+			f.lggr.AssumptionViolationw("invalid receipt found for confirmed transaction", "receipt", receipt)
+			continue
+		}
+		// The tx store query only returns transactions with receipts older than or equal to the finalized block but leaving this check here for a belts and braces approach
+		if receipt.BlockNumber > latestFinalizedHead.BlockNumber() {
+			continue
+		}
+		// Receipt block num older than earliest head in chain. Validate hash using RPC call later
+		if receipt.BlockNumber < earliestBlockNumInChain {
+			blockNumToReceiptsMap[receipt.BlockNumber] = append(blockNumToReceiptsMap[receipt.BlockNumber], receipt)
+			continue
+		}
+		blockHashInChain := latestFinalizedHead.HashAtHeight(receipt.BlockNumber)
+		// Receipt block hash does not match the block hash in chain. Transaction has been re-org'd out but DB state has not been updated yet
+		if blockHashInChain.String() != receipt.BlockHash.String() {
+			// Log error if a transaction is marked as confirmed with a receipt older than the finalized block
+			// This scenario could potentially point to a re-org'd transaction the Confirmer has lost track of
+			f.lggr.Errorw("found confirmed transaction with re-org'd receipt older than finalized block", "receipt", receipt, "onchainBlockHash", blockHashInChain.String())
+			continue
+		}
+		finalizedReceipts = append(finalizedReceipts, receipt)
+	}
+
+	// Check if block hashes exist for receipts on-chain older than the earliest cached head
+	// Transactions are grouped by their receipt block hash to avoid repeat requests on the same hash in case transactions were confirmed in the same block
+	validatedReceipts, err := f.batchCheckReceiptHashesOnchain(ctx, blockNumToReceiptsMap)
+	if err != nil {
+		// Do not error out to allow transactions that did not need RPC validation to still be marked as finalized
+		// The transactions failed to be validated will be checked again in the next round
+		f.lggr.Errorf("failed to validate receipt block hashes over RPC: %v", err)
+	}
+	finalizedReceipts = append(finalizedReceipts, validatedReceipts...)
+
+	receiptIDs := f.buildReceiptIdList(finalizedReceipts)
+
+	err = f.txStore.UpdateTxStatesToFinalizedUsingReceiptIds(ctx, receiptIDs, f.chainId)
+	if err != nil {
+		return fmt.Errorf("failed to update transactions as finalized: %w", err)
+	}
+	// Update lastProcessedFinalizedBlockNum after processing has completed to allow failed processing to retry on subsequent heads
+	// Does not need to be protected with mutex lock because the Finalizer only runs in a single loop
+	f.lastProcessedFinalizedBlockNum = latestFinalizedHead.BlockNumber()
+	return nil
+}
+
+func (f *evmFinalizer) batchCheckReceiptHashesOnchain(ctx context.Context, blockNumToReceiptsMap map[int64][]Receipt) ([]Receipt, error) {
```

Comment:
This method doesn't seem to return an error in any path. You can probably drop it.

Reply:
Good catch! It's removed in the latest commit